// @storybook/vue/dist/server/framework-preset-vue.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.webpack = webpack;
exports.babelDefault = babelDefault;

var _plugin = _interopRequireDefault(require("vue-loader/lib/plugin"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function webpack(config) {
  return _objectSpread({}, config, {
    plugins: [...config.plugins, new _plugin.default()],
    module: _objectSpread({}, config.module, {
      rules: [...config.module.rules, {
        test: /\.vue$/,
        loader: require.resolve('vue-loader'),
        options: {}
      }]
    }),
    resolve: _objectSpread({}, config.resolve, {
      extensions: [...config.resolve.extensions, '.vue'],
      alias: _objectSpread({}, config.resolve.alias, {
        vue$: require.resolve('vue/dist/vue.esm.js')
      })
    })
  });
}

function babelDefault(config) {
  return _objectSpread({}, config, {
    presets: [...config.presets, require.resolve('babel-preset-vue')]
  });
}
